### ISSUE TITLE
Updating the UTXO set at each block

### ISSUE THREAD

**Issue by nelsontodd** [citp/BlockSci#108]
I work at [Unchained Capital](https://www.unchained-capital.com) and I'm trying to recreate a copy of our [UTXO Age Distribution](https://blog.unchained-capital.com/bitcoin-data-science-pt-1-hodl-waves-7f3501d53f63).

I need to iterate over the UTXO set at each block and group the UTXOs by their age.  I see the [outputsUnspent](https://github.com/citp/BlockSci/blob/master/blockscipy/src/chain/output/output_range_filters.cpp#L22) method of the `Blockchain` class but this only allows iteration over the *current* UTXO set, and not historical snapshots of it.

I see in [your paper](https://arxiv.org/pdf/1709.02489.pdf) (Fig. 13) that you calculate a timeseries of the percentage BTC unspent in the previous month -- how did you use BlockSci to extract this data? If you could share sample code that would be great.

I understand why caching the UTXO set at each block is not a default behavior of BlockSci (lots of space!).  I would like to be able to turn on this behavior if possible.

**Comment by hkalodner** [comment]
Hi @nelsontodd,

I thought recreating this graph was an interesting challenge so I took a swing at doing this myself. For queries such as these I've found the optimal approach is to reduce the problem into calculating the net change per block rather than defining it on the UTXO pool at each height. I've written some code previously that produces a rolling UTXO pool object that can advance blocks and changes, but I haven't found a compelling use case for this feature.

Some of it is a bit convoluted and I there certainly might be some minor bugs, but the linked gist shows a solution to this problem which runs in a little under 3 minutes on our r4.2xlarge EC2 instance.

https://gist.github.com/hkalodner/b389a5009969de6b6141031829c9640b

Here's the graph which it generates. It looks like it matches up nicely with your own.

![utxo_age](https://user-images.githubusercontent.com/412311/39741335-866dd6fa-5267-11e8-98f7-26ccafa4097f.png)
